catalog/descs: hydrate types for tuples

yuzefovich opened the pull request:
Before we can use the user-defined types, we have to hydrate them. We do
so for enums as well as arrays of enums, but we forgot to do so for
tuples containing enums in some cases. This is now fixed.

Release note (bug fix): Previously, CockroachDB could encounter an
internal error or crash when some queries involving tuples with ENUMs
were executed in a distributed manner. This is now fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72406)
<!-- Reviewable:end -->

yuzefovich commented:
I confirmed that the reproduction crashes on both 21.2 and 21.1 with `fakedist` configs, so we should backport this to those releases.

postamar reviewed (changes requested):
This change would solve the immediate problem you're facing, but it appears we also have a deeper problem that won't go away when it comes to type hydration in general. Part of what prompted this is that we now have a limited form of record types: implicit table record types.

From what I can tell this logic you've lifted in `maybeHydrateType` appears in many different places around where `HydrateTypeInfoWithName` is called. Probably because what we _really_ want is to make sure a `*types.T` is properly recursively hydrated, we don't really care about the type descriptors per se.

I'm arguing that you should lift it further into `typedesc` :
1. Its definition should be something like `func EnsureTypeIsHydrated(ctx context.Context, t *types.T, res catalog.TypeDescriptorResolver) error` or whichever name you prefer.
2. It should recursively hydrate types just like `HydrateTypeSlice` and `HydrateTypesInTableDescriptor` do.
3. Since nobody will ever really want to call `HydrateTypeInfoWithName` directly this method should be package-private, or subsumed by `EnsureTypeIsHydrated` somehow.

If we don't do this, it seems like this class of bug will just keep reappearing 😬

yuzefovich committed:
catalog/descs: hydrate types for tuples

Before we can use the user-defined types, we have to hydrate them. We do
so for enums as well as arrays of enums, but we forgot to do so for
tuples containing enums in some cases. This is now fixed.

Release note (bug fix): Previously, CockroachDB could encounter an
internal error or crash when some queries involving tuples with ENUMs
were executed in a distributed manner. This is now fixed.

yuzefovich reviewed (commented):
Thanks for the suggestions! I implemented points 1 and 2, but I don't know how we would unexport `HydrateTypeInfoWithName` - `TypeDescriptor` interface is defined in `catalog` package, but its implementations live in `typedesc`. `EnsureTypeIsHydrated` is agnostic of those implementations, so I don't think it can subsume `HydrateTypeInfoWithName`. Am I missing something?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72406#-:-Mnh8ezq4sMICtGRY91Z:bivginf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
LGTM, thanks for doing this.

What I meant is that with your change, the call sites for the `HydrateTypeInfoWithName` method are all in the `typedesc` package so there's no point any more in having it declared as part of the `catalog.TypeDescriptor`, nor making it public. My gut feeling is that there's not a lot of potential for future usage of that method either. It's really not important, I just like to be quite aggressive in tightening APIs, but that's just me.

yuzefovich commented:
Yeah, I understand your thinking, but I don't know how to achieve that technically - like we have an interface that lives in one package, and then a couple of implementations of that interface living in another package.

We cannot make that method private in the interface, so we'd need to remove the method from `TypeDescriptor` interface. We probably would then introduce a separate `typeDescriptor` interface into `typedesc` package something like:
```
type typeDescriptor {
    catalog.TypeDescriptor
    hydrateTypeInfoWithName func(...) error
}
```
then in `EnsureTypeIsHydrated` we'd have to add an interface conversion with something like
```
return elemTypDesc.(typeDescriptor).HydrateTypeInfoWithName(ctx, t, &elemTypName, res)
```
because `GetTypeDescriptor` returns `catalog.TypeDescriptor` and is used in other places.

I guess I could make a change like this. Thoughts? Do you see another way of unexporting `HydrateTypeInfoWithName`?

yuzefovich commented:
Implemented that idea in a separate commit, let me know what you think (I'll squash it if we decide to keep it).

yuzefovich commented:
I guess personally I'm probably against this change because if we were to introduce another implementation of `catalog.TypeDescriptor`, there are no compile time checks that would ensure that implementation also implements `typedesc.typeDescriptor`.

postamar commented:
Oh shoot I forgot about `TableImplicitRecordType`. Never mind, sorry for wasting your time.

yuzefovich commented:
No worries, TFTR!

bors r+
